File: Python/Algorithms/BubbleSort.py
# Regular bubble sort implementation
def bubbleSort(arr):
    size = len(arr)
    for i in range(size - 1):
        for j in range(size - i - 1):
            if arr[j] > arr[j + 1]:
                swap(arr, j, j + 1)

# A faster version of bubble sort!
def bubbleSortFaster(arr):
    size = len(arr)
    for i in range(size - 1):
        swapped = False
        for j in range(size - i - 1):
            if  arr[j] > arr[j + 1]:
                swap(arr,  j, j + 1)
                swapped = True
        
        if not swapped:
            break

# Helper function to swap elements in place
def swap(arr, a, b):
    tmp = arr[a]
    arr[a] = arr[b]
    arr[b] = tmp

# ----------
# Demo Code
# ----------
array = [2,5,3,7,9,4,6]
print(bubbleSort(array.copy()))
print(bubbleSortFaster(array.copy()))

File: CPlusPlus/Algorithms/BubbleSort.cpp
#include <stdio.h>

void sort(int arr[], int size) {

    for(int i = 0; i < size - 1; ++i) {
        for(int j = 0; j < size - i - 1; ++j) {

            if(arr[j] > arr[j+1]) {
                int tmp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = tmp;
            }
        }
    }
}

// Faster version of sort
void sortFaster(int arr[], int size) {
    bool swapped;
    for(int i = 0; i < size - 1; ++i) {
        swapped = false;
        for(int j = 0; j < size - i - 1; ++j) {
            if(arr[j] > arr[j+1]) {
                int tmp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = tmp;

                swapped = true;
            }
        }
        if(!swapped) {
            break;
        }
    }
}

// Demo:
int main() {
    // Test array:
    int a[] = {4, 8, 5, 1, 2, 0, 9};
    // Get the number of elements in the array:
    int s = sizeof(a) / sizeof(a[0]);

    // Sort the array:
    sort(a, s);

    for(int i = 0; i < s; ++i) {
        printf("%d ", a[i]);
    }
    printf("\n");

    return 0;
}